From 72f1c82b29fe0af41fad4696a9ed7576d510ccf3 Mon Sep 17 00:00:00 2001
From: Chen Zhongjin <chenzhongjin@huawei.com>
Date: Wed, 8 Feb 2023 22:53:04 -0500
Subject: [PATCH 882/905] [Huawei] samples: bpf: Add sample BMC for Redis

Offering: HULK
hulk inclusion
category: feature
bugzilla: N/A

--------------------------------

BMC is an in-kernel key-value cache implemented in BPF and proposed by
paper [1]. The paper discussed BMC for memcached, obtaining at least
6x performance speedup.

This patch implements a sample BMC for Redis.

See [2] for details on how to build samples/bpf.

Output files:
 samples/bpf/bmctool
 samples/bpf/bmc/bpf.o

Sample usage:
 bmctool prog load -p 6379 ./bmc/bpf.o  # load bmc bpf prog and attach it
                                        # to sockets with listen port 6379

 bmctool stat                           # dump bmc status

 bmctool prog unload                    # detach and unload bmc prog

Tested with the following command:

 ./redis-benchmark -c 20 -r 1 -n 1000 -t get  -h 192.168.4.101 -d 102

Without BMC:
  throughput summary: 41666.67 requests per second
  latency summary (msec):
          avg       min       p50       p95       p99       max
        0.441     0.176     0.415     0.631     1.455     1.815

With BMC (100% HIT):
  throughput summary: 66666.67 requests per second
  latency summary (msec):
          avg       min       p50       p95       p99       max
        0.223     0.096     0.215     0.311     0.743     0.759

BMC Stat:
 Total GET Requests: 1000
 Hit GET Requests: 1000 (100.00%)
 Dropped GET Requests: 0 (0.00%)
 Total SET Requests: 1
 Hit SET Requests: 1 (100.00%)
 Dropped SET Requests: 0 (0.00%)

[1] https://www.usenix.org/conference/nsdi21/presentation/ghigoff
[2] https://www.kernel.org/doc/readme/samples-bpf-README.rst

Signed-off-by: Chen Zhongjin <chenzhongjin@huawei.com>
Signed-off-by: Xu Kuohai <xukuohai@huawei.com>
Signed-off-by: Yang Jihong <yangjihong@huawei.com>
Signed-off-by: He Fengqing <hefengqing@huawei.com> (original demo)
---
 samples/bpf/Makefile     |    3 +
 samples/bpf/bmc/bpf.c    |  792 ++++++++++++++++++++++++++++++
 samples/bpf/bmc/common.h |   25 +
 samples/bpf/bmc/tool.c   | 1000 ++++++++++++++++++++++++++++++++++++++
 4 files changed, 1820 insertions(+)
 create mode 100644 samples/bpf/bmc/bpf.c
 create mode 100644 samples/bpf/bmc/common.h
 create mode 100644 samples/bpf/bmc/tool.c

diff --git a/samples/bpf/Makefile b/samples/bpf/Makefile
index aeebf5d12f32..639767d587a1 100644
--- a/samples/bpf/Makefile
+++ b/samples/bpf/Makefile
@@ -54,6 +54,7 @@ tprogs-y += task_fd_query
 tprogs-y += xdp_sample_pkts
 tprogs-y += ibumad
 tprogs-y += hbm
+tprogs-y := bmctool
 
 # Libbpf dependencies
 LIBBPF = $(TOOLS_PATH)/lib/bpf/libbpf.a
@@ -111,6 +112,7 @@ task_fd_query-objs := bpf_load.o task_fd_query_user.o $(TRACE_HELPERS)
 xdp_sample_pkts-objs := xdp_sample_pkts_user.o $(TRACE_HELPERS)
 ibumad-objs := bpf_load.o ibumad_user.o $(TRACE_HELPERS)
 hbm-objs := bpf_load.o hbm.o $(CGROUP_HELPERS)
+bmctool-objs := bmc/tool.o
 
 # Tell kbuild to always build the programs
 always-y := $(tprogs-y)
@@ -172,6 +174,7 @@ always-y += ibumad_kern.o
 always-y += hbm_out_kern.o
 always-y += hbm_edt_kern.o
 always-y += xdpsock_kern.o
+always-y += bmc/bpf.o
 
 ifeq ($(ARCH), arm)
 # Strip all except -D__LINUX_ARM_ARCH__ option needed to handle linux
diff --git a/samples/bpf/bmc/bpf.c b/samples/bpf/bmc/bpf.c
new file mode 100644
index 000000000000..88cc4859dd66
--- /dev/null
+++ b/samples/bpf/bmc/bpf.c
@@ -0,0 +1,792 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright (C) Huawei Technologies Co., Ltd. 2022-2022. All rights reserved.
+ *
+ * Description: BPF program to accelerate Redis. The idea is to add a kernel
+ * cache for Redis data. When new Redis request is received, the kernel cache
+ * is checked, and if the requested data is found in the cache, a Redis reply
+ * message is constructed and sent back directly.
+ */
+
+#include <uapi/linux/in.h>
+#include <uapi/linux/if_ether.h>
+#include <uapi/linux/ip.h>
+#include <uapi/linux/tcp.h>
+#include <uapi/linux/bpf.h>
+#include <uapi/linux/pkt_cls.h>
+
+#include <bpf/bpf_endian.h>
+#include <bpf/bpf_helpers.h>
+
+#include "common.h"
+
+#define BMC_MAX_REDIS_KEY_LEN	256
+#define BMC_MAX_REDIS_VALUE_LEN 512
+#define BMC_MAX_REDIS_HEAD_SLEN	4
+#define BMC_MAX_REDIS_EX_SLEN	7	// max len - ms per sec len
+#define BMC_REDIS_KEY_SIZE		(BMC_MAX_REDIS_KEY_LEN + 16)
+#define BMC_REDIS_VALUE_SIZE	(BMC_MAX_REDIS_VALUE_LEN + 16)
+#define BMC_REDIS_EXP_SIZE		(BMC_MAX_REDIS_EX_SLEN + 16)
+#define BMC_MAX_TCP_SIZE (sizeof(struct tcphdr) + BMC_REDIS_VALUE_SIZE)
+
+#define ETH_HDR(data)	((struct ethhdr *)data)
+#define IP_HDR(data)	((struct iphdr *)(ETH_HDR(data) + 1))
+#define TCP_HDR(data)	((struct tcphdr *)(IP_HDR(data) + 1))
+#define TCP_BODY(data)	((char *)(TCP_HDR(data) + 1))
+
+#define IP_MF		0x2000
+#define IP_OFFSET	0x1FFF
+#define NS_PER_MS	1000000
+#define MS_PER_SEC	1000
+#define ktime_get_ms()	(bpf_ktime_get_ns() / NS_PER_MS)
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PROG_ARRAY);
+	__uint(key_size, sizeof(u32));
+	__uint(value_size, sizeof(u32));
+	__uint(max_entries, 3);
+} bmc_jump_table SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(key_size, sizeof(u32));
+	__uint(value_size, sizeof(u32));
+	__uint(max_entries, 16);
+} bmc_ports SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_ARRAY);
+	__uint(key_size, sizeof(u32));
+	__uint(value_size, sizeof(u32));
+	__uint(max_entries, 1);
+} bmc_interface SEC(".maps");
+
+struct redis_key {
+	u32 len;
+	/* encoded in redis format */
+	u8 data[BMC_REDIS_KEY_SIZE];
+};
+
+struct redis_value {
+	u32 len;
+	u64 expiry;
+	/* encoded in redis format */
+	u8 data[BMC_REDIS_VALUE_SIZE];
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_LRU_HASH);
+	__uint(key_size, sizeof(struct redis_key));
+	__uint(value_size, sizeof(struct redis_value));
+	__uint(max_entries, 100000);
+} bmc_storage SEC(".maps");
+
+struct redis_ctx {
+	struct redis_key key;
+	struct redis_value value;
+	u32 expiry_len;
+	u32 expiry_value;
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+	__uint(key_size, sizeof(u32));
+	__uint(value_size, sizeof(struct redis_ctx));
+	__uint(max_entries, 1);
+} ctxmap SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+	__uint(key_size, sizeof(u32));
+	__uint(value_size, sizeof(struct redis_bmc_stat));
+	__uint(max_entries, 1);
+} bmc_stats SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(key_size, 16);
+	__uint(value_size, sizeof(u64));
+	__uint(max_entries, 1024);
+} bmc_cmds SEC(".maps");
+
+static inline struct redis_ctx *get_ctx(void)
+{
+	u32 key = 0;
+
+	return bpf_map_lookup_elem(&ctxmap, &key);
+}
+
+static inline struct redis_bmc_stat *get_stat(void)
+{
+	u32 key = 0;
+
+	return bpf_map_lookup_elem(&bmc_stats, &key);
+}
+
+static bool is_bmc_port(u32 port)
+{
+	u32 *val = bpf_map_lookup_elem(&bmc_ports, &port);
+
+	return val != NULL && *val != 0;
+}
+
+static inline void set_tcp_tuple(struct bpf_sock_tuple *tuple, char *data, u32 len)
+{
+	struct iphdr *ihdr = IP_HDR(data);
+	struct tcphdr *thdr = TCP_HDR(data);
+
+	tuple->ipv4.saddr = ihdr->saddr;
+	tuple->ipv4.daddr = ihdr->daddr;
+	tuple->ipv4.sport = thdr->source;
+	tuple->ipv4.dport = thdr->dest;
+
+	tuple->ipv4.seq = __bpf_ntohl(thdr->seq);
+	tuple->ipv4.rcv_len = __bpf_ntohs(ihdr->tot_len) -
+						   sizeof(struct iphdr) - thdr->doff * 4;
+	tuple->ipv4.ack_seq = __bpf_ntohl(thdr->ack_seq);
+	tuple->ipv4.snd_len = len;
+}
+
+/*
+ * Call both bpf_xdp_adjust_tail and bpf_update_tcp_seq safely.
+ * If any of them failed, this return -1 and caller should return XDP_PASS.
+ */
+static int prepare_tcp(struct xdp_md *xdp, char *data, char *data_end, u32 len)
+{
+	struct bpf_sock_tuple tuple;
+	int old_len = data_end - data;
+	int new_len = (TCP_BODY(data) - data) + len;
+
+	set_tcp_tuple(&tuple, data, len);
+
+	if (old_len >= new_len) {
+		if (bpf_update_tcp_seq(xdp, &tuple, sizeof(tuple.ipv4), -1, 0))
+			return -1;
+
+		/* adjust won't fail when shrink xdp size */
+		bpf_xdp_adjust_tail(xdp, new_len - old_len);
+	} else {
+		if (bpf_xdp_adjust_tail(xdp, new_len - old_len))
+			return -1;
+
+		if (bpf_update_tcp_seq(xdp, &tuple, sizeof(tuple.ipv4), -1, 0)) {
+			/* shrink xdp size to rollback the buff */
+			bpf_xdp_adjust_tail(xdp, old_len - new_len);
+			return -1;
+		}
+	}
+
+	return 0;
+}
+
+/* initialize protocal headers, the data ptr must be valid */
+static inline void build_reply_head(char *data, u32 len)
+{
+	struct ethhdr *ehdr = ETH_HDR(data);
+	struct iphdr *ihdr = IP_HDR(data);
+	struct tcphdr *thdr = TCP_HDR(data);
+	unsigned int seq;
+
+	seq = __bpf_ntohl(thdr->seq) + __bpf_ntohs(ihdr->tot_len) -
+		  sizeof(struct iphdr) - thdr->doff * 4;
+	thdr->seq = thdr->ack_seq;
+	thdr->ack_seq = __bpf_ntohl(seq);
+	thdr->doff = 5; /* discard tcp options */
+	ihdr->tot_len = __bpf_htons(sizeof(struct iphdr) +
+					sizeof(struct tcphdr) + len);
+
+	swap(thdr->source, thdr->dest);
+	swap(ihdr->saddr, ihdr->daddr);
+	for (size_t i = 0; i < ETH_ALEN; i++)
+		swap(ehdr->h_source[i], ehdr->h_dest[i]);
+}
+
+static inline bool is_redis_get(char *s)
+{
+	return s[0] == '$' && s[1] == '3' && s[2] == '\r' && s[3] == '\n' &&
+		   ((s[4] == 'g' && s[5] == 'e' && s[6] == 't') ||
+			(s[4] == 'G' && s[5] == 'E' && s[6] == 'T')) &&
+		   s[7] == '\r' && s[8] == '\n';
+}
+
+static inline bool is_redis_set(char *s)
+{
+	return s[0] == '$' && s[1] == '3' && s[2] == '\r' && s[3] == '\n' &&
+		   ((s[4] == 's' && s[5] == 'e' && s[6] == 't') ||
+			(s[4] == 'S' && s[5] == 'E' && s[6] == 'T')) &&
+		   s[7] == '\r' && s[8] == '\n';
+}
+
+static inline bool is_redis_setex(char *s)
+{
+	return s[0] == '$' && s[1] == '5' && s[2] == '\r' && s[3] == '\n' &&
+		   ((s[4] == 's' && s[5] == 'e' && s[6] == 't' && s[7] == 'e' && s[8] == 'x') ||
+		   (s[4] == 'S' && s[5] == 'E' && s[6] == 'T' && s[7] == 'E' && s[8] == 'X')) &&
+		   s[9] == '\r' && s[10] == '\n';
+}
+
+static inline void record_redis_cmd(char *payload)
+{
+	u64 val = 0;
+	u64 *pval = NULL;
+
+	pval = bpf_map_lookup_elem(&bmc_cmds, payload);
+	if (pval != NULL)
+		val = *pval;
+	val++;
+	bpf_map_update_elem(&bmc_cmds, payload, &val, BPF_ANY);
+}
+
+static inline int xdp_store_str(char *dst, char *src, u32 len, char *data_end)
+{
+	if (dst + len > data_end)
+		return -1;
+
+	memcpy(dst, src, len);
+	return 0;
+}
+
+static u32 xdp_load_bytes(char *dst, char *src, u32 len, u32 max_len, char *data_end)
+{
+	size_t i, j;
+
+	/* copy qwords */
+	unsigned long *dstl = (unsigned long *)dst;
+	unsigned long *srcl = (unsigned long *)src;
+
+	for (i = 0; i < max_len && i + sizeof(unsigned long) <= len;
+		 i += sizeof(unsigned long)) {
+		if (srcl + 1 > data_end)
+			return 0;
+		*dstl++ = *srcl++;
+	}
+
+	/* copy left bytes */
+	char *dstb = (char *)dstl;
+	char *srcb = (char *)srcl;
+
+	for (j = 0; j < sizeof(unsigned long) &&
+		 i + j < max_len && i + j < len; j++) {
+		if (srcb + 1 > data_end)
+			return 0;
+		*dstb++ = *srcb++;
+	}
+
+	return i + j;
+}
+
+static u32 xdp_store_bytes(char *dst, char *src, u32 len, u32 max_len, char *data_end)
+{
+	size_t i, j;
+
+	/* copy qwords */
+	unsigned long *dstl = (unsigned long *)dst;
+	unsigned long *srcl = (unsigned long *)src;
+
+	for (i = 0; i < max_len && i + sizeof(unsigned long) <= len;
+		 i += sizeof(unsigned long)) {
+		if (dstl + 1 > data_end)
+			return 0;
+		*dstl++ = *srcl++;
+	}
+
+	/* copy left bytes */
+	char *dstb = (char *)dstl;
+	char *srcb = (char *)srcl;
+
+	for (j = 0; j < sizeof(unsigned long) &&
+		 i + j < max_len && i + j < len; j++) {
+		if (dstb + 1 > data_end)
+			return 0;
+		*dstb++ = *srcb++;
+	}
+
+	return i + j;
+}
+
+static char *extract_kvdata(char *payload, int max_size, char *data_end)
+{
+	char *kv_data = payload;
+	u32 kv_len = 0;
+
+	if (payload >= data_end || payload[0] != '$')
+		return NULL;
+	payload++;
+
+	/* size head */
+	for (size_t i = 0; i < BMC_MAX_REDIS_HEAD_SLEN; i++) {
+		if (payload >= data_end || payload[0] < '0' || payload[0] > '9')
+			break;
+		kv_len = kv_len * 10 + (payload[0] - '0');
+		payload++;
+	}
+	if (payload + 2 > data_end || payload[0] != '\r' || payload[1] != '\n')
+		return NULL;
+	payload += 2;
+
+	/* data body */
+	payload += kv_len;
+	if (payload + 2 > data_end ||
+		payload[0] != '\r' || payload[1] != '\n')
+		return NULL;
+	payload += 2;
+
+	if (payload - kv_data > max_size)
+		return NULL;
+	return payload;
+}
+
+static char *extract_exdata(char *payload, u32 *ex_value, char *data_end)
+{
+	char *kv_data = payload;
+	u32 len = 0;
+	u32 val = 0;
+
+	if (payload >= data_end || payload[0] != '$')
+		return NULL;
+	payload++;
+
+	for (size_t i = 0; i < BMC_MAX_REDIS_HEAD_SLEN; i++) {
+		if (payload >= data_end || payload[0] < '0' || payload[0] > '9')
+			break;
+		len = len * 10 + (payload[0] - '0');
+		payload++;
+	}
+	if (payload + 2 > data_end || payload[0] != '\r' || payload[1] != '\n')
+		return NULL;
+	payload += 2;
+
+	for (size_t i = 0; i < BMC_MAX_REDIS_EX_SLEN && i < len; i++) {
+		if (payload >= data_end || payload[0] < '0' || payload[0] > '9')
+			break;
+		val = val * 10 + (payload[0] - '0');
+		payload++;
+	}
+	if (payload + 2 > data_end ||
+		payload[0] != '\r' || payload[1] != '\n')
+		return NULL;
+	payload += 2;
+
+	*ex_value = val;
+	return payload;
+}
+
+static inline void ip_csum(struct iphdr *ihdr, char *data_end)
+{
+	u32 csum = 0;
+	u16 *next_ip_u16 = (u16 *)ihdr;
+
+	ihdr->check = 0;
+	for (size_t i = 0; i < sizeof(struct iphdr); i += 2) {
+		if (next_ip_u16 + 1 > data_end)
+			break;
+		csum += *next_ip_u16++;
+	}
+
+	csum = (u16)csum + (csum >> 16);
+	csum += csum >> 16;
+	ihdr->check = (u16) ~csum;
+}
+
+static inline void tcp_csum(struct iphdr *ihdr, struct tcphdr *thdr, char *data_end)
+{
+	u32 csum = 0;
+	u16 *next_ip_u16 = (u16 *)thdr;
+
+	thdr->check = 0;
+	csum += (u16)ihdr->saddr;
+	csum += (u16)(ihdr->saddr >> 16);
+	csum += (u16)ihdr->daddr;
+	csum += (u16)(ihdr->daddr >> 16);
+	csum += (u16)ihdr->protocol << 8;
+	csum += bpf_htons(bpf_ntohs(ihdr->tot_len) - ihdr->ihl * 4);
+
+	for (size_t i = 0; i < BMC_MAX_TCP_SIZE; i += 2) {
+		if (next_ip_u16 + 1 > data_end)
+			break;
+		csum += *next_ip_u16++;
+	}
+
+	if ((char *)next_ip_u16 + 1 <= data_end) {
+		u8 r[2] = { *(u8 *)next_ip_u16, 0 };
+
+		csum += *(u16 *)r;
+	}
+
+	csum = (u16)csum + (csum >> 16);
+	csum += csum >> 16;
+	thdr->check = (u16) ~csum;
+}
+
+/* update checksum for xdp, the data ptr must be valid */
+static void v4_csum(char *data, char *data_end)
+{
+	ip_csum(IP_HDR(data), data_end);
+	tcp_csum(IP_HDR(data), TCP_HDR(data), data_end);
+}
+
+SEC("bmc/get")
+int bmc_get(struct xdp_md *xdp)
+{
+	u32 vlen;
+	struct redis_ctx *ctx;
+	struct redis_value *val;
+	struct redis_bmc_stat *stat;
+
+	char *data = (char *)(long)xdp->data;
+	char *data_end = (char *)(long)xdp->data_end;
+	char reply[] = { '$', '-', '1', '\r', '\n'};
+	char *payload, *dst;
+
+	if (TCP_BODY(data) > data_end)
+		return XDP_PASS;
+
+	ctx = get_ctx();
+	if (!ctx)
+		return XDP_PASS;
+
+	stat = get_stat();
+	if (!stat)
+		return XDP_PASS;
+
+	stat->total_get_requests++;
+
+	payload = (void *)TCP_HDR(data) +
+			  TCP_HDR(data)->doff * 4; // tcp head offset
+	/* skip "*2\r\n$3\r\nget\r\n" */
+	payload += 13;
+
+	memset(&ctx->key.data, 0, sizeof(u8) * BMC_REDIS_KEY_SIZE);
+	xdp_load_bytes(ctx->key.data, payload, ctx->key.len, BMC_REDIS_KEY_SIZE, data_end);
+
+	val = bpf_map_lookup_elem(&bmc_storage, &ctx->key);
+
+	/* pass to userspace */
+	if (!val || !val->len || val->len > BMC_REDIS_VALUE_SIZE)
+		return XDP_PASS;
+
+	/*
+	 * expired, just reply -1
+	 * don't remove element in map, it can be used for next hit
+	 */
+	if (val->expiry && ktime_get_ms() > val->expiry) {
+		if (prepare_tcp(xdp, data, data_end, sizeof(reply)))
+			return XDP_PASS;
+
+		data = (char *)(long)xdp->data;
+		data_end = (char *)(long)xdp->data_end;
+		if (TCP_BODY(data) > data_end)
+			goto drop;
+
+		build_reply_head(data, sizeof(reply));
+
+		dst = TCP_BODY(data);
+		if (xdp_store_str(dst, reply, sizeof(reply), data_end))
+			goto drop;
+
+		v4_csum(data, data_end);
+
+		stat->hit_get_requests++;
+		return XDP_TX;
+	}
+
+	vlen = val->len;
+
+	if (prepare_tcp(xdp, data, data_end, vlen))
+		return XDP_PASS;
+
+	data = (char *)(long)xdp->data;
+	data_end = (char *)(long)xdp->data_end;
+	if (TCP_BODY(data) > data_end)
+		goto drop;
+
+	build_reply_head(data, vlen);
+
+	dst = TCP_BODY(data);
+	xdp_store_bytes(dst, val->data, vlen, BMC_REDIS_VALUE_SIZE, data_end);
+
+	v4_csum(data, data_end);
+
+	stat->hit_get_requests++;
+	return XDP_TX;
+drop:
+	stat->drop_get_requests++;
+	return XDP_DROP;
+
+}
+
+SEC("bmc/set")
+int bmc_set(struct xdp_md *xdp)
+{
+	struct redis_ctx *ctx;
+	struct redis_value *val;
+	struct redis_bmc_stat *stat;
+
+	char *data = (char *)(long)xdp->data;
+	char *data_end = (char *)(long)xdp->data_end;
+	char reply[] = { '+', 'O', 'K', '\r', '\n'};
+	char *payload, *dst;
+
+	if (TCP_BODY(data) > data_end)
+		return XDP_PASS;
+
+	ctx = get_ctx();
+	if (!ctx)
+		return XDP_PASS;
+
+	stat = get_stat();
+	if (!stat)
+		return XDP_PASS;
+
+	stat->total_set_requests++;
+
+	payload = (void *)TCP_HDR(data) +
+			  TCP_HDR(data)->doff * 4; // tcp head offset
+	/* skip "*2\r\n$3\r\nset\r\n" */
+	payload += 13;
+
+	memset(&ctx->key.data, 0, sizeof(u8) * BMC_REDIS_KEY_SIZE);
+	xdp_load_bytes(ctx->key.data, payload, ctx->key.len, BMC_REDIS_KEY_SIZE, data_end);
+
+	if (ctx->key.len > BMC_REDIS_KEY_SIZE)
+		return XDP_PASS;
+	payload += ctx->key.len;
+	xdp_load_bytes(ctx->value.data, payload, ctx->value.len, BMC_REDIS_VALUE_SIZE, data_end);
+
+	ctx->value.expiry = 0;
+
+	if (bpf_map_update_elem(&bmc_storage, &ctx->key,
+							&ctx->value, BPF_ANY))
+		return XDP_PASS;
+
+	if (prepare_tcp(xdp, data, data_end, sizeof(reply)))
+		return XDP_PASS;
+
+	data = (char *)(long)xdp->data;
+	data_end = (char *)(long)xdp->data_end;
+	if (TCP_BODY(data) > data_end)
+		goto drop;
+
+	build_reply_head(data, sizeof(reply));
+
+	dst = TCP_BODY(data);
+	if (xdp_store_str(dst, reply, sizeof(reply), data_end))
+		goto drop;
+
+	v4_csum(data, data_end);
+
+	stat->hit_set_requests++;
+	return XDP_TX;
+drop:
+	stat->drop_set_requests++;
+	return XDP_DROP;
+}
+
+SEC("bmc/setex")
+int bmc_setex(struct xdp_md *xdp)
+{
+	struct redis_ctx *ctx;
+	struct redis_value *val;
+	struct redis_bmc_stat *stat;
+
+	char *data = (char *)(long)xdp->data;
+	char *data_end = (char *)(long)xdp->data_end;
+	char reply[] = { '+', 'O', 'K', '\r', '\n'};
+	char *payload, *dst;
+
+	if (TCP_BODY(data) > data_end)
+		return XDP_PASS;
+
+	ctx = get_ctx();
+	if (!ctx)
+		return XDP_PASS;
+
+	stat = get_stat();
+	if (!stat)
+		return XDP_PASS;
+
+	stat->total_setex_requests++;
+
+	payload = (void *)TCP_HDR(data) +
+			  TCP_HDR(data)->doff * 4; // tcp head offset
+	/* skip "*4\r\n$5\r\nsetex\r\n" */
+	payload += 15;
+
+	memset(&ctx->key.data, 0, sizeof(u8) * BMC_REDIS_KEY_SIZE);
+	xdp_load_bytes(ctx->key.data, payload, ctx->key.len, BMC_REDIS_KEY_SIZE, data_end);
+
+	ctx->value.expiry = ktime_get_ms() + (u64)ctx->expiry_value * MS_PER_SEC;
+
+	if (ctx->key.len > BMC_REDIS_KEY_SIZE ||
+		ctx->expiry_len > BMC_REDIS_EXP_SIZE)
+		return XDP_PASS;
+	payload += ctx->key.len + ctx->expiry_len;
+	xdp_load_bytes(ctx->value.data, payload, ctx->value.len, BMC_REDIS_VALUE_SIZE, data_end);
+
+	if (bpf_map_update_elem(&bmc_storage, &ctx->key,
+							&ctx->value, BPF_ANY))
+		return XDP_PASS;
+
+	if (prepare_tcp(xdp, data, data_end, sizeof(reply)))
+		return XDP_PASS;
+
+	data = (char *)(long)xdp->data;
+	data_end = (char *)(long)xdp->data_end;
+	if (TCP_BODY(data) > data_end)
+		goto drop;
+
+	build_reply_head(data, sizeof(reply));
+
+	dst = TCP_BODY(data);
+	if (xdp_store_str(dst, reply, sizeof(reply), data_end))
+		goto drop;
+
+	v4_csum(data, data_end);
+
+	stat->hit_setex_requests++;
+	return XDP_TX;
+drop:
+	stat->drop_setex_requests++;
+	return XDP_DROP;
+}
+
+/*
+ * GET message format:
+ * "*2\r\n"	// this is an array with 3 elements
+ * "$3\r\n"	// the first element is a string with 3 characters
+ * "get\r\n"	// the string is "get"
+ * "$5\r\n"	// the second element is a string with 5 characters
+ * "key01\r\n"	// the string is "key01"
+ *
+ * SET message format:
+ * "*3\r\n"	// this is an array with 3 elements
+ * "$3\r\n"	// the first element is a string with 3 characters
+ * "set\r\n"	// the string is "set"
+ * "$5\r\n"	// the second element is a string with 5 characters
+ * "key01\r\n"	// the string is "key01"
+ * "$5\r\n"	// the third element is a string with 5 characters
+ * "val01\r\n"  // the string is "valu01"
+ *
+ * SETEX message format:
+ * "*4\r\n"	// this is an array with 4 elements
+ * "$5\r\n"	// the first element is a string with 5 characters
+ * "setex\r\n"	// the string is "setex"
+ * "$5\r\n"	// the second element is a string with 5 characters
+ * "key01\r\n"	// the string is "key01"
+ * "$2\r\n"	// the third element is a string with 2 characters
+ * "60\r\n"	// the string is "60"
+ * "$5\r\n"	// the forth element is a string with 5 characters
+ * "val01\r\n"  // the string is "valu01"
+ */
+SEC("bmc/main")
+int bmc_main(struct xdp_md *xdp)
+{
+	char *data = (char *)(long)xdp->data;
+	char *data_end = (char *)(long)xdp->data_end;
+	struct ethhdr *ehdr = NULL;
+	struct iphdr *ihdr = NULL;
+	struct tcphdr *thdr = NULL;
+	struct redis_ctx *ctx;
+	char *payload, *start;
+
+	ehdr = ETH_HDR(data);
+	if (ehdr + 1 > data_end)
+		return XDP_PASS;
+
+	if (ehdr->h_proto != __bpf_constant_htons(ETH_P_IP))
+		return XDP_PASS;
+
+	ihdr = IP_HDR(data);
+	if (ihdr + 1 > data_end)
+		return XDP_PASS;
+
+	/* NO IP options support */
+	if (ihdr->ihl != 5 || ihdr->protocol != IPPROTO_TCP)
+		return XDP_PASS;
+
+	if (ihdr->frag_off & __bpf_htons(IP_MF | IP_OFFSET))
+		return XDP_PASS;
+
+	if (__bpf_htons(ihdr->tot_len) > ETH_DATA_LEN)
+		return XDP_PASS;
+
+	thdr = TCP_HDR(data);
+	if (thdr + 1 > data_end)
+		return XDP_PASS;
+
+	if (thdr->syn || thdr->fin || thdr->rst)
+		return XDP_PASS;
+
+	if (!is_bmc_port(thdr->dest))
+		return XDP_PASS;
+
+	payload = (void *)TCP_HDR(data) +
+			  TCP_HDR(data)->doff * 4; // tcp head offset
+	if (payload + 20 > data_end)
+		return XDP_PASS;
+
+	if (payload[0] != '*' || payload[2] != '\r' || payload[3] != '\n')
+		return XDP_PASS;
+
+	ctx = get_ctx();
+	if (!ctx)
+		return XDP_PASS;
+
+	switch (payload[1]) {
+	case '2':
+		/* skip "*2\r\n" */
+		if (is_redis_get(payload + 4)) {
+			/* skip "*2\r\n$3\r\nget\r\n" */
+			payload += 13;
+			if (extract_kvdata(payload, BMC_REDIS_KEY_SIZE, data_end) != data_end)
+				return XDP_PASS;
+			ctx->key.len = data_end - payload;
+			bpf_tail_call(xdp, &bmc_jump_table, REDIS_GET_PROG_INDEX);
+		}
+		break;
+	case '3':
+		/* skip "*3\r\n" */
+		if (is_redis_set(payload + 4)) {
+			/* skip "*2\r\n$3\r\nset\r\n" */
+			payload += 13;
+			start = payload;
+			payload = extract_kvdata(payload, BMC_REDIS_KEY_SIZE, data_end);
+			if (!payload)
+				return XDP_PASS;
+			ctx->key.len = payload - start;
+
+			if (extract_kvdata(payload, BMC_REDIS_VALUE_SIZE, data_end) != data_end)
+				return XDP_PASS;
+			ctx->value.len = data_end - payload;
+			bpf_tail_call(xdp, &bmc_jump_table, REDIS_SET_PROG_INDEX);
+		}
+		break;
+	case '4':
+		/* skip "*4\r\n" */
+		if (is_redis_setex(payload + 4)) {
+			/* skip "*4\r\n$5\r\nsetex\r\n" */
+			payload += 15;
+			start = payload;
+			payload = extract_kvdata(payload, BMC_REDIS_KEY_SIZE, data_end);
+			if (!payload)
+				return XDP_PASS;
+			ctx->key.len = payload - start;
+
+			start = payload;
+			payload = extract_exdata(payload, &ctx->expiry_value, data_end);
+			if (!payload)
+				return XDP_PASS;
+			ctx->expiry_len = payload - start;
+
+			if (extract_kvdata(payload, BMC_REDIS_VALUE_SIZE, data_end) != data_end)
+				return XDP_PASS;
+			ctx->value.len = data_end - payload;
+			bpf_tail_call(xdp, &bmc_jump_table, REDIS_SETEX_PROG_INDEX);
+		}
+		break;
+	}
+
+	return XDP_PASS;
+}
+
+char _license[] SEC("license") = "GPL";
diff --git a/samples/bpf/bmc/common.h b/samples/bpf/bmc/common.h
new file mode 100644
index 000000000000..4abda2effd91
--- /dev/null
+++ b/samples/bpf/bmc/common.h
@@ -0,0 +1,25 @@
+/* SPDX-License-Identifier: GPL-2.0
+ *
+ * Copyright (C) Huawei Technologies Co., Ltd. 2022-2022. All rights reserved.
+ * Description: common header for both user prog and bpf kernel prog
+ */
+#ifndef __REDIS_BMC_COMMON_H__
+#define __REDIS_BMC_COMMON_H__
+
+#define REDIS_GET_PROG_INDEX	0
+#define REDIS_SET_PROG_INDEX	1
+#define REDIS_SETEX_PROG_INDEX	2
+
+struct redis_bmc_stat {
+	__u64 total_get_requests;
+	__u64 hit_get_requests;
+	__u64 drop_get_requests;
+	__u64 total_set_requests;
+	__u64 hit_set_requests;
+	__u64 drop_set_requests;
+	__u64 total_setex_requests;
+	__u64 hit_setex_requests;
+	__u64 drop_setex_requests;
+};
+
+#endif
diff --git a/samples/bpf/bmc/tool.c b/samples/bpf/bmc/tool.c
new file mode 100644
index 000000000000..c9496eaf4b17
--- /dev/null
+++ b/samples/bpf/bmc/tool.c
@@ -0,0 +1,1000 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright (C) Huawei Technologies Co., Ltd. 2022-2022. All rights reserved.
+ */
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <limits.h>
+#include <errno.h>
+
+#include <linux/if_link.h>
+
+#include <unistd.h>
+#include <net/if.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/select.h>
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include <arpa/inet.h>
+#include <fcntl.h>
+
+#include <bpf/bpf.h>
+#include <bpf/libbpf.h>
+
+#include "common.h"
+
+#define DEFAULT_CGROUP_PATH "/sys/fs/cgroup"
+#define DEFAULT_REDIS_PORT  6379
+#define ARRAY_SIZE(x) (sizeof(x) / sizeof((x)[0]))
+#define IFINDEX_NUM 8
+
+struct {
+	char *cgroup_path;
+	char *bpf_path;
+	int cgroup_fd;
+	int map_jump_table_fd;
+	int map_ports_fd;
+	int map_storage_fd;
+	int map_interface_fd;
+	int map_stats_fd;
+	int map_cmds_fd;
+	int redis_main_prog_fd;
+	int redis_get_prog_fd;
+	int redis_set_prog_fd;
+	int redis_setex_prog_fd;
+	int xdp_mode;
+	uint16_t listen_port;
+	unsigned int ifindex;
+} bmc;
+
+struct bmc_prog_info {
+	const char *sec_name;
+	enum bpf_prog_type prog_type;
+	enum bpf_attach_type attach_type;
+	int *p_prog_fd;
+	int *p_attach_fd;
+	unsigned int attach_flags;
+	unsigned int is_xdp_main;
+	const char *pin_path;
+	struct bpf_program *prog;
+};
+
+struct bmc_map_info {
+	const char *map_name;
+	int *p_map_fd;
+	char *pin_path;
+	struct bpf_map *map;
+	bool is_stat_map;
+	bool is_interface_map;
+};
+
+static struct bmc_prog_info prog_infos[] = {
+	{
+		.sec_name = "bmc/main",
+		.prog_type = BPF_PROG_TYPE_XDP,
+		.p_prog_fd = &bmc.redis_main_prog_fd,
+		.attach_flags = XDP_FLAGS_DRV_MODE, // XDP_FLAGS_SKB_MODE
+		.is_xdp_main = 1,
+		.pin_path = "/sys/fs/bpf/bmc/prog_bmc_main"
+	},
+	{
+		.sec_name = "bmc/get",
+		.prog_type = BPF_PROG_TYPE_XDP,
+		.p_prog_fd = &bmc.redis_get_prog_fd,
+		.p_attach_fd = NULL,
+		.attach_flags = 0,
+		.pin_path = "/sys/fs/bpf/bmc/prog_bmc_redis_get"
+	},
+	{
+		.sec_name = "bmc/set",
+		.prog_type = BPF_PROG_TYPE_XDP,
+		.p_prog_fd = &bmc.redis_set_prog_fd,
+		.p_attach_fd = NULL,
+		.attach_flags = 0,
+		.pin_path = "/sys/fs/bpf/bmc/prog_bmc_redis_set"
+	},
+	{
+		.sec_name = "bmc/setex",
+		.prog_type = BPF_PROG_TYPE_XDP,
+		.p_prog_fd = &bmc.redis_setex_prog_fd,
+		.p_attach_fd = NULL,
+		.attach_flags = 0,
+		.pin_path = "/sys/fs/bpf/bmc/prog_bmc_redis_setex"
+	},
+};
+
+static struct bmc_map_info map_infos[] = {
+	{
+		.map_name = "bmc_jump_table",
+		.p_map_fd = &bmc.map_jump_table_fd,
+		.pin_path = "/sys/fs/bpf/bmc/map_jump_table"
+	},
+	{
+		.map_name = "bmc_ports",
+		.p_map_fd = &bmc.map_ports_fd,
+		.pin_path = "/sys/fs/bpf/bmc/map_ports"
+	},
+	{
+		.map_name = "bmc_storage",
+		.p_map_fd = &bmc.map_storage_fd,
+		.pin_path = "/sys/fs/bpf/bmc/map_storage"
+	},
+	{
+		.map_name = "bmc_interface",
+		.p_map_fd = &bmc.map_interface_fd,
+		.pin_path = "/sys/fs/bpf/bmc/interface",
+		.is_interface_map = true,
+	},
+	{
+		.map_name = "bmc_stats",
+		.p_map_fd = &bmc.map_stats_fd,
+		.pin_path = "/sys/fs/bpf/bmc/stats",
+		.is_stat_map = true,
+	},
+	{
+		.map_name = "bmc_cmds",
+		.p_map_fd = &bmc.map_cmds_fd,
+		.pin_path = "/sys/fs/bpf/bmc/cmds",
+	},
+};
+
+static int find_type_by_sec_name(const char *sec_name,
+				 enum bpf_prog_type *p_prog_type,
+				 enum bpf_attach_type *p_attach_type)
+{
+	int i;
+
+	if (sec_name == NULL) {
+		fprintf(stderr, "sec_name is NULL\n");
+		return -1;
+	}
+
+	for (i = 0; i < ARRAY_SIZE(prog_infos); i++) {
+		if (!strcmp(prog_infos[i].sec_name, sec_name)) {
+			*p_prog_type = prog_infos[i].prog_type;
+			*p_attach_type = prog_infos[i].attach_type;
+			return 0;
+		}
+	}
+
+	fprintf(stderr, "unknown prog %s\n", sec_name);
+
+	return -1;
+}
+
+static int set_prog_type(struct bpf_object *obj)
+{
+	const char *sec_name;
+	struct bpf_program *prog;
+	enum bpf_prog_type prog_type;
+	enum bpf_attach_type attach_type;
+
+	bpf_object__for_each_program(prog, obj) {
+		sec_name = bpf_program__section_name(prog);
+		if (find_type_by_sec_name(sec_name, &prog_type, &attach_type))
+			return -1;
+		bpf_program__set_type(prog, prog_type);
+		if (prog_type != BPF_PROG_TYPE_XDP)
+			bpf_program__set_expected_attach_type(prog, attach_type);
+	}
+
+	return 0;
+}
+
+static struct bpf_object *load_bpf_file(const char *bpf_file)
+{
+	int err;
+	char err_buf[256];
+	struct bpf_object *obj;
+
+	obj = bpf_object__open(bpf_file);
+	err = libbpf_get_error(obj);
+	if (err) {
+		libbpf_strerror(err, err_buf, sizeof(err_buf));
+		fprintf(stderr, "unable to open bpf file %s : %s\n", bpf_file,
+			err_buf);
+		return NULL;
+	}
+
+	if (set_prog_type(obj)) {
+		bpf_object__close(obj);
+		return NULL;
+	}
+
+	err = bpf_object__load(obj);
+	if (err) {
+		fprintf(stderr, "load bpf object failed\n");
+		bpf_object__close(obj);
+		return NULL;
+	}
+
+	return obj;
+}
+
+static int find_prog(struct bpf_object *obj, const char *sec_name,
+		     struct bpf_program **p_prog, int *p_prog_fd)
+{
+	int fd;
+	struct bpf_program *prog;
+
+	prog = bpf_object__find_program_by_title(obj, sec_name);
+	if (!prog) {
+		fprintf(stderr, "failed to find prog %s\n", sec_name);
+		return -1;
+	}
+
+	fd = bpf_program__fd(prog);
+	if (fd < 0) {
+		fprintf(stderr, "failed to get fd of prog %s\n", sec_name);
+		return -1;
+	}
+
+	*p_prog = prog;
+	*p_prog_fd = fd;
+
+	return 0;
+}
+
+static void unpin_progs(int n)
+{
+	int i;
+
+	for (i = 0; i < n; i++)
+		bpf_program__unpin(prog_infos[i].prog, prog_infos[i].pin_path);
+}
+
+static int find_progs(struct bpf_object *obj)
+{
+	int i;
+	struct bmc_prog_info *info;
+
+	for (i = 0; i < ARRAY_SIZE(prog_infos); i++) {
+		info = &prog_infos[i];
+
+		if (find_prog(obj, info->sec_name, &info->prog, info->p_prog_fd))
+			goto error_find_prog;
+
+		if (bpf_program__pin(info->prog, info->pin_path))
+			goto error_find_prog;
+	}
+
+	return 0;
+
+error_find_prog:
+	unpin_progs(i);
+	return -1;
+}
+
+static int find_map(struct bpf_object *obj, const char *map_name,
+		    struct bpf_map **p_map, int *p_map_fd)
+{
+	int fd;
+	struct bpf_map *map;
+
+	map = bpf_object__find_map_by_name(obj, map_name);
+	if (!map) {
+		fprintf(stderr, "failed to find map %s\n", map_name);
+		return -1;
+	}
+
+	fd = bpf_map__fd(map);
+	if (fd < 0) {
+		fprintf(stderr, "failed to get fd of map %s\n", map_name);
+		return -1;
+	}
+
+
+	*p_map = map;
+	*p_map_fd = fd;
+
+	return 0;
+}
+
+static void unpin_maps(int n)
+{
+	int i;
+
+	for (i = 0; i < n; i++)
+		bpf_map__unpin(map_infos[i].map, map_infos[i].pin_path);
+}
+
+static int find_maps(struct bpf_object *obj)
+{
+	int i;
+	__u32 key;
+	__u32 value;
+	int fd;
+	struct bmc_map_info *info;
+
+	for (i = 0; i < ARRAY_SIZE(map_infos); i++) {
+		info = &map_infos[i];
+
+		if (find_map(obj, info->map_name, &info->map, info->p_map_fd))
+			goto error_find_map;
+
+		if (bpf_map__pin(info->map, info->pin_path)) {
+			fprintf(stderr, "failed to pin map %s to path %s\n",
+				info->map_name, info->pin_path);
+			goto error_find_map;
+		}
+
+		if (info->is_interface_map) {
+			key = 0;
+			value = bmc.ifindex;
+			fd = bpf_map__fd(info->map);
+			bpf_map_update_elem(fd, &key, &value, 0);
+		}
+	}
+
+	return 0;
+
+error_find_map:
+	unpin_maps(i);
+	return -1;
+}
+
+static void detach_xdp_progs(unsigned int ifindex, __u32 flags)
+{
+	if (bmc.xdp_mode == XDP_FLAGS_SKB_MODE)
+		flags = XDP_FLAGS_SKB_MODE;
+	else if (bmc.xdp_mode == XDP_FLAGS_DRV_MODE)
+		flags = XDP_FLAGS_DRV_MODE;
+
+	bpf_set_link_xdp_fd(ifindex, -1, flags);
+}
+
+static void detach_progs(int n)
+{
+	int i;
+	struct bmc_prog_info *info;
+
+	for (i = 0; i < n; i++) {
+		info = &prog_infos[i];
+		if (info->is_xdp_main)
+			detach_xdp_progs(bmc.ifindex, info->attach_flags);
+		else if (info->prog_type != BPF_PROG_TYPE_XDP)
+			bpf_prog_detach(*info->p_prog_fd, info->attach_type);
+	}
+}
+
+static int attach_xdp_prog(int prog_fd, __u32 flags)
+{
+	if (bmc.xdp_mode == XDP_FLAGS_SKB_MODE)
+		flags = XDP_FLAGS_SKB_MODE;
+	else if (bmc.xdp_mode == XDP_FLAGS_DRV_MODE)
+		flags = XDP_FLAGS_DRV_MODE;
+
+	if (bmc.ifindex) {
+		if (bpf_set_link_xdp_fd(bmc.ifindex, prog_fd, flags)) {
+			fprintf(stderr, "failed to attach xdp prog\n");
+			return  -1;
+		}
+	}
+	return 0;
+}
+
+static int attach_progs(struct bpf_object *obj)
+{
+	int i;
+	int err;
+	int prog_fd;
+	int attach_fd;
+	unsigned int flags;
+	enum bpf_attach_type type;
+	struct bmc_prog_info *info;
+
+	for (i = 0; i < ARRAY_SIZE(prog_infos); i++) {
+		info = &prog_infos[i];
+		prog_fd = *info->p_prog_fd;
+		flags = info->attach_flags;
+
+		if (info->is_xdp_main)
+			err = attach_xdp_prog(prog_fd, flags);
+		else if (info->prog_type != BPF_PROG_TYPE_XDP &&
+			 info->p_attach_fd != NULL) {
+			attach_fd = *info->p_attach_fd;
+			type = info->attach_type;
+			err = bpf_prog_attach(prog_fd, attach_fd, type, flags);
+		} else
+			continue;
+
+		if (err) {
+			fprintf(stderr, "attach prog %s failed!\n",
+				info->sec_name);
+			goto error_attach_prog;
+		}
+	}
+
+	return 0;
+
+error_attach_prog:
+	detach_progs(i);
+
+	return -1;
+}
+
+static int add_bmc_port(void)
+{
+	int ret;
+	int map_fd = bmc.map_ports_fd;
+	uint16_t port = htons(bmc.listen_port);
+	uint32_t key = (uint32_t)port;
+	uint32_t value = 1;
+
+	ret = bpf_map_update_elem(map_fd, &key, &value, 0);
+	if (ret)
+		fprintf(stderr, "failed to add port %u\n", port);
+
+	return ret;
+}
+
+static int add_tail_call(void)
+{
+	int ret;
+	int map_fd = bmc.map_jump_table_fd;
+	__u32 key;
+	__u32 value;
+
+	key = REDIS_GET_PROG_INDEX;
+	value = bmc.redis_get_prog_fd;
+	ret = bpf_map_update_elem(map_fd, &key, &value, 0);
+	if (ret) {
+		fprintf(stderr, "failed to add redis get tail call prog\n");
+		return -1;
+	}
+
+	key = REDIS_SET_PROG_INDEX;
+	value = bmc.redis_set_prog_fd;
+	ret = bpf_map_update_elem(map_fd, &key, &value, 0);
+	if (ret) {
+		fprintf(stderr, "failed to add redis set tail call prog\n");
+		key = REDIS_GET_PROG_INDEX;
+		bpf_map_delete_elem(map_fd, &key);
+	}
+
+	key = REDIS_SETEX_PROG_INDEX;
+	value = bmc.redis_setex_prog_fd;
+	ret = bpf_map_update_elem(map_fd, &key, &value, 0);
+	if (ret) {
+		fprintf(stderr, "failed to add redis setex tail call prog\n");
+		key = REDIS_SET_PROG_INDEX;
+		bpf_map_delete_elem(map_fd, &key);
+		key = REDIS_GET_PROG_INDEX;
+		bpf_map_delete_elem(map_fd, &key);
+	}
+
+	return ret;
+}
+
+static int setup_bpf(void)
+{
+	struct bpf_object *obj;
+
+	bmc.cgroup_fd = open(bmc.cgroup_path, O_DIRECTORY, O_RDONLY);
+	if (bmc.cgroup_fd < 0) {
+		fprintf(stderr, "failed to open cgroup %s: %s\n",
+			bmc.cgroup_path, strerror(errno));
+		return -1;
+	}
+
+	obj = load_bpf_file(bmc.bpf_path);
+	if (!obj)
+		goto error_load_object;
+
+	if (find_progs(obj))
+		goto error_load_object;
+
+	if (find_maps(obj))
+		goto error_find_maps;
+
+	if (attach_progs(obj))
+		goto error_attach_progs;
+
+	if (add_bmc_port())
+		goto error_add_port;
+
+	if (add_tail_call())
+		goto error_attach_progs;
+
+	return 0;
+
+error_add_port:
+	detach_progs(ARRAY_SIZE(prog_infos));
+error_attach_progs:
+	unpin_maps(ARRAY_SIZE(map_infos));
+error_find_maps:
+	unpin_progs(ARRAY_SIZE(prog_infos));
+error_load_object:
+	bpf_object__close(obj);
+	close(bmc.cgroup_fd);
+	return -1;
+}
+
+static int parse_load_args(int argc, char *argv[])
+{
+	int opt;
+	int port;
+	const char *ifname = NULL;
+
+	bmc.cgroup_path = DEFAULT_CGROUP_PATH;
+	bmc.listen_port = DEFAULT_REDIS_PORT;
+	bmc.ifindex = 0;
+	bmc.xdp_mode = 0;
+
+	while ((opt = getopt(argc, argv, "c:p:i:m:")) != -1) {
+		switch (opt) {
+		case 'c':
+			bmc.cgroup_path = optarg;
+			break;
+		case 'p':
+			port = atoi(optarg);
+			if (port <= 0 || port >= USHRT_MAX) {
+				fprintf(stderr, "invalid port: %s\n", optarg);
+				return -1;
+			}
+			bmc.listen_port = port;
+			break;
+		case 'i':
+			printf("interface: %s\n", optarg);
+			ifname = optarg;
+			bmc.ifindex = if_nametoindex(ifname);
+			break;
+		case 'm':
+			if (!strcmp(optarg, "generic")) {
+				bmc.xdp_mode = XDP_FLAGS_SKB_MODE;
+			} else if (!strcmp(optarg, "native")) {
+				bmc.xdp_mode = XDP_FLAGS_DRV_MODE;
+			} else {
+				fprintf(stderr, "invalid mode: %s, ", optarg);
+				fprintf(stderr, "only generic or native is allowed\n");
+				return -1;
+			}
+			break;
+		default:
+			fprintf(stderr, "unknown option %c\n", opt);
+			return -1;
+		}
+	}
+
+	if (!bmc.ifindex) {
+		fprintf(stderr, "no netwrok interface found\n");
+		return -1;
+	}
+
+	if (optind >= argc) {
+		fprintf(stderr, "no bpf prog file found\n");
+		return -1;
+	}
+
+	bmc.bpf_path = argv[optind];
+
+	printf("bpf file: %s\n", bmc.bpf_path);
+	printf("cgroup path: %s\n", bmc.cgroup_path);
+	printf("listen port: %d\n", bmc.listen_port);
+	printf("interface: %s\n", ifname);
+
+	return 0;
+}
+
+static int parse_unload_args(int argc, char *argv[])
+{
+	int opt;
+
+	bmc.cgroup_path = DEFAULT_CGROUP_PATH;
+	bmc.xdp_mode = 0;
+
+	while ((opt = getopt(argc, argv, "c:m:")) != -1) {
+		switch (opt) {
+		case 'c':
+			bmc.cgroup_path = optarg;
+			break;
+		case 'm':
+			if (!strcmp(optarg, "generic")) {
+				bmc.xdp_mode = XDP_FLAGS_SKB_MODE;
+			} else if (!strcmp(optarg, "native")) {
+				bmc.xdp_mode = XDP_FLAGS_DRV_MODE;
+			} else {
+				fprintf(stderr, "invalid mode: %s, ", optarg);
+				fprintf(stderr, "only generic or native is allowed\n");
+				return -1;
+			}
+			break;
+		default:
+			fprintf(stderr, "unknown option %c\n", opt);
+			return -1;
+		}
+	}
+
+	return 0;
+}
+
+struct cmd {
+	const char *name;
+	int (*func)(int argc, char *argv[]);
+};
+
+static int do_prog(int argc, char *argv[]);
+static int do_stat(int argc, char *argv[]);
+
+static int do_prog_load(int argc, char *argv[]);
+static int do_prog_unload(int argc, char *argv[]);
+
+static struct cmd main_cmds[] = {
+	{ "prog", do_prog },
+	{ "stat", do_stat },
+};
+
+static struct cmd prog_cmds[] = {
+	{ "load", do_prog_load },
+	{ "unload", do_prog_unload },
+};
+
+static char *elf_name;
+
+static int dispatch_cmd(struct cmd cmds[], int ncmd, int argc,
+			char *argv[], void (*help)(void))
+{
+	int i;
+	int ret;
+
+	if (argc <= 0) {
+		help();
+		return -1;
+	}
+
+	for (i = 0; i < ncmd; i++) {
+		if (!strcmp(argv[0], cmds[i].name)) {
+			ret = cmds[i].func(argc - 1, argv + 1);
+			if (ret == -2) {
+				help();
+				ret = -1;
+			}
+			return ret;
+		}
+	}
+
+	help();
+
+	return -1;
+}
+
+static int do_prog_load(int argc, char *argv[])
+{
+	if (parse_load_args(argc + 1, argv - 1) < 0)
+		return -2;
+
+	if (setup_bpf())
+		return -1;
+
+	return 0;
+}
+
+static int do_prog_unload(int argc, char *argv[])
+{
+	int i;
+	int err;
+	int prog_fd;
+	int cgroup_fd;
+	int map_fd;
+	char *interface_map_path = NULL;
+	__u32 ifindex;
+	__u32 key;
+
+	if (parse_unload_args(argc + 1, argv - 1) < 0)
+		return -2;
+
+	cgroup_fd = open(bmc.cgroup_path, O_DIRECTORY, O_RDONLY);
+	if (cgroup_fd < 0) {
+		fprintf(stderr, "failed to open cgroup path: %s\n",
+			bmc.cgroup_path);
+		return -1;
+	}
+
+	for (i = 0; i < ARRAY_SIZE(map_infos); i++) {
+		if (map_infos[i].is_interface_map) {
+			interface_map_path = map_infos[i].pin_path;
+			break;
+		}
+	}
+
+	if (!interface_map_path) {
+		fprintf(stderr, "no interface map found\n");
+		return -1;
+	}
+
+	map_fd = bpf_obj_get(interface_map_path);
+	if (map_fd < 0) {
+		fprintf(stderr, "failed to get map from %s\n",
+			interface_map_path);
+		return -1;
+	}
+
+	key = 0;
+	err = bpf_map_lookup_elem(map_fd, &key, &ifindex);
+	close(map_fd);
+	if (err) {
+		fprintf(stderr, "lookup interface failed\n");
+		return -1;
+	}
+
+	for (i = 0; i < ARRAY_SIZE(prog_infos); i++) {
+		if (prog_infos[i].attach_type == BPF_CGROUP_SOCK_OPS) {
+			prog_fd = bpf_obj_get(prog_infos[i].pin_path);
+			if (prog_fd >= 0)
+				bpf_prog_detach2(prog_fd, cgroup_fd,
+						BPF_CGROUP_SOCK_OPS);
+		}
+
+		if (prog_infos[i].is_xdp_main)
+			detach_xdp_progs(ifindex, prog_infos[i].attach_flags);
+
+		unlink(prog_infos[i].pin_path);
+	}
+
+	for (i = 0; i < ARRAY_SIZE(map_infos); i++)
+		unlink(map_infos[i].pin_path);
+
+	return 0;
+}
+
+static void do_prog_help(void)
+{
+	fprintf(stderr, "Usage: %s prog load\n", elf_name);
+	fprintf(stderr, "		[-c CGROUP_PATH] [-p LISTEN_PORT]\n");
+	fprintf(stderr, "		{-i INTERFACE} {BPF_FILE}\n");
+	fprintf(stderr, "       %s prog unload [CGROUP_PATH]\n", elf_name);
+}
+
+static int do_prog(int argc, char *argv[])
+{
+	return dispatch_cmd(prog_cmds, ARRAY_SIZE(prog_cmds),
+			    argc, argv, do_prog_help);
+}
+
+static int do_list_cmds(int argc, char *argv[])
+{
+	int i;
+	int fd;
+	int err;
+	bool found = false;
+	struct bmc_map_info *info;
+	struct bpf_map_info map = {};
+	__u32 len = sizeof(map);
+	char *key;
+	char next_key[32] = { 0 };
+
+	for (i = 0; i < ARRAY_SIZE(map_infos); i++) {
+		info = &map_infos[i];
+		if (!strcmp(info->map_name, "bmc_cmds")) {
+			found = true;
+			break;
+		}
+	}
+
+	if (!found)
+		return -1;
+
+	fd = bpf_obj_get(info->pin_path);
+	if (fd < 0) {
+		fprintf(stderr, "failed to open %s\n",
+			info->pin_path);
+		return -1;
+	}
+
+	err = bpf_obj_get_info_by_fd(fd, &map, &len);
+	if (err) {
+		fprintf(stderr, "failed to get map info\n");
+		err = -1;
+		goto out;
+	}
+
+
+	key = NULL;
+	while (bpf_map_get_next_key(fd, key, next_key) > -1) {
+		char data[32];
+		char *p = data;
+		char *cmd = NULL;
+		__u64 count;
+
+		key = next_key;
+
+		found = false;
+		memcpy(data, next_key, 32);
+		for (i = 0; i < 31; i++) {
+			if (p[i] == '\r' && p[i + 1] == '\n') {
+				found = true;
+				i += 2;
+				break;
+			}
+		}
+
+		if (!found)
+			continue;
+
+		found = false;
+		while (i < 31) {
+			if (p[i] == '\r' && p[i + 1] == '\n') {
+				found = true;
+				i += 2;
+				cmd = p + i;
+				break;
+			}
+			i++;
+		}
+
+		if (!found)
+			continue;
+
+		while (i < 31) {
+			if (p[i] == '\r' && p[i + 1] == '\n')
+				break;
+			i++;
+		}
+
+		p[i] = 0;
+
+		bpf_map_lookup_elem(fd, next_key, &count);
+
+		printf("cmd: %s, count=%lld\n", cmd, count);
+	}
+out:
+	close(fd);
+	return err;
+}
+
+static int do_stat(int argc, char *argv[])
+{
+	int i;
+	int fd;
+	int err;
+	int ncpu;
+	bool found = false;
+	struct bmc_map_info *info;
+	struct bpf_map_info map = {};
+	struct redis_bmc_stat *percpu_stat;
+	struct redis_bmc_stat stat = {};
+	__u32 len = sizeof(map);
+	__u32 key;
+
+	do_list_cmds(argc, argv);
+
+	ncpu = sysconf(_SC_NPROCESSORS_ONLN);
+	if (ncpu < 0) {
+		fprintf(stderr, "sysconf failed: %s\n", strerror(errno));
+		return -1;
+	}
+
+	percpu_stat = malloc(sizeof(struct redis_bmc_stat) * ncpu);
+	if (!percpu_stat) {
+		fprintf(stderr, "malloc percpu stat failed\n");
+		return -1;
+	}
+
+	for (i = 0; i < ARRAY_SIZE(map_infos); i++) {
+		info = &map_infos[i];
+		if (info->is_stat_map) {
+			found = true;
+			break;
+		}
+	}
+
+	if (!found) {
+		fprintf(stderr, "no stats map found\n");
+		free(percpu_stat);
+		return -1;
+	}
+
+	fd = bpf_obj_get(info->pin_path);
+	if (fd < 0) {
+		fprintf(stderr, "failed to open %s\n",
+			info->pin_path);
+		free(percpu_stat);
+		return -1;
+	}
+
+	err = bpf_obj_get_info_by_fd(fd, &map, &len);
+	if (err) {
+		fprintf(stderr, "failed to get map info\n");
+		err = -1;
+		goto out;
+	}
+
+	if (map.type != BPF_MAP_TYPE_PERCPU_ARRAY) {
+		fprintf(stderr, "unexpected map type: %d\n", map.type);
+		err = -1;
+		goto out;
+	}
+
+	if (map.key_size != sizeof(__u32)) {
+		fprintf(stderr, "unexpected map key_size: %u\n", map.key_size);
+		err = -1;
+		goto out;
+	}
+
+	if (map.value_size != sizeof(struct redis_bmc_stat)) {
+		fprintf(stderr, "unexpected map key_size: %u\n", map.key_size);
+		err = -1;
+		goto out;
+	}
+
+	key = 0;
+	err = bpf_map_lookup_elem(fd, &key, percpu_stat);
+	if (err) {
+		fprintf(stderr, "lookup cpu stat failed, cpu=%u\n", i);
+		err = -1;
+		goto out;
+	}
+
+	for (int i = 0; i < ncpu; i++) {
+		stat.total_get_requests += percpu_stat[i].total_get_requests;
+		stat.hit_get_requests += percpu_stat[i].hit_get_requests;
+		stat.drop_get_requests += percpu_stat[i].drop_get_requests;
+		stat.total_set_requests += percpu_stat[i].total_set_requests;
+		stat.hit_set_requests += percpu_stat[i].hit_set_requests;
+		stat.drop_set_requests += percpu_stat[i].drop_set_requests;
+		stat.total_setex_requests += percpu_stat[i].total_setex_requests;
+		stat.hit_setex_requests += percpu_stat[i].hit_setex_requests;
+		stat.drop_setex_requests += percpu_stat[i].drop_setex_requests;
+	}
+
+	printf("Total GET Requests: %llu\n", stat.total_get_requests);
+	printf("Hit GET Requests: %llu (%.2f%%)\n", stat.hit_get_requests,
+		stat.total_get_requests == 0 ? 0 :
+		(double)stat.hit_get_requests /
+		(double)stat.total_get_requests *
+		100);
+	printf("Dropped GET Requests: %llu (%.2lf%%)\n", stat.drop_get_requests,
+		stat.total_get_requests == 0 ? 0 :
+		(double)stat.drop_get_requests /
+		(double)stat.total_get_requests *
+		100);
+
+	printf("Total SET Requests: %llu\n", stat.total_set_requests);
+	printf("Hit SET Requests: %llu (%.2f%%)\n", stat.hit_set_requests,
+		stat.total_set_requests == 0 ? 0 :
+		(double)stat.hit_set_requests /
+		(double)stat.total_set_requests *
+		100);
+	printf("Dropped SET Requests: %llu (%.2lf%%)\n", stat.drop_set_requests,
+		stat.total_set_requests == 0 ? 0 :
+		(double)stat.drop_set_requests /
+		(double)stat.total_set_requests *
+		100);
+
+	printf("Total SETEX Requests: %llu\n", stat.total_setex_requests);
+	printf("Hit SETEX Requests: %llu (%.2f%%)\n", stat.hit_setex_requests,
+		stat.total_setex_requests == 0 ? 0 :
+		(double)stat.hit_setex_requests /
+		(double)stat.total_setex_requests *
+		100);
+	printf("Dropped SETEX Requests: %llu (%.2lf%%)\n", stat.drop_setex_requests,
+		stat.total_setex_requests == 0 ? 0 :
+		(double)stat.drop_setex_requests /
+		(double)stat.total_setex_requests *
+		100);
+
+out:
+	close(fd);
+	free(percpu_stat);
+
+	return err;
+}
+
+static void do_main_help(void)
+{
+	fprintf(stderr,
+		"Usage: %s OBJECT { COMMAND | help }\n"
+		"       OBJECT := { prog | stat }\n",
+		elf_name);
+}
+
+int main(int argc, char *argv[])
+{
+	elf_name = argv[0];
+
+	return dispatch_cmd(main_cmds, ARRAY_SIZE(main_cmds),
+			    argc - 1, argv + 1, do_main_help);
+}
-- 
2.17.1

